### ISSUE TITLE
Upgrade Java Version

### ISSUE THREAD

**Issue by lorenzleutgeb** [alpha-asp/Alpha#301]
Migrate Alpha to [Java 17](https://openjdk.java.net/projects/jdk/17/).

[GraalVM 21.3](https://www.graalvm.org/release-notes/21_3/) was released on 2021-10-19 and supports Java 17.

Blockers:
 - #274 
 - https://github.com/gradle/gradle/issues/16857

**Comment by lorenzleutgeb** [comment]
Gradle 7.3 with Support for Java 17 is out: https://docs.gradle.org/7.3/release-notes.html

**Comment by lorenzleutgeb** [comment]
When migrating, make sure that Checkstyle behaves as expected early on.

**Comment by lorenzleutgeb** [comment]
Now that #274 is merged, this is unblocked.

@AntoniusW how do you feel about raising the minimum required JDK version to 17? It's [available for Ubuntu](https://packages.ubuntu.com/search?suite=default&section=all&arch=amd64&keywords=openjdk-17-jdk&searchon=names) in all current versions ([see Ubuntu version history on Wikipedia](https://en.wikipedia.org/wiki/Ubuntu_version_history#Version_timeline)), and since it is the current Java LTS release (see [Java version history on Wikipedia](https://en.wikipedia.org/wiki/Java_version_history)) it will be supported until 2029. Alpha skipped one LTS release, namely Java 11, in between.

Changing the minimum required JDK version to Java 17, does have some implications regarding dependencies using the [Java Platform Module System (JPMS)](https://en.wikipedia.org/wiki/Java_Platform_Module_System) which was introduced in Java 9. There are no backward incompatible changes to the syntax, so in terms of the "the real code", i.e. our `*.java` files, I don't expect many changes.

**Comment by AntoniusW** [comment]
TBH, I think it is an advantage if our software is able to run on "ancient" versions of Java. I don't mind switching CI to include all available and modern versions of Java, but requiring it without some good reasons feels like unnecessary self-constraining. The problem is, once we switch to a higher version, it will get harder and harder to switch back, because of course someone will try out new features. If we decide that the new features are really worth it, then I am also in favor of switching to more modern version, but so far, I did not see those features we really want to have.

You name JPMS, which is not available in Java 8 and which -- for some reasons, I guess -- is not widely adopted. I am simply not convinced so far, that JPMS really is something we want to do and that it is worth the requirement of a higher Java version.
In other words, what are the benefits of requiring Java 17?

**Comment by lorenzleutgeb** [comment]
> TBH, I think it is an advantage if our software is able to run on "ancient" versions of Java. I don't mind switching CI to include all available and modern versions of Java, but requiring it without some good reasons feels like unnecessary self-constraining.
> The problem is, once we switch to a higher version, it will get harder and harder to switch back, because of course someone will try out new features.

Why would you switch back? Java 17 is stable.

> If we decide that the new features are really worth it, then I am also in favor of switching to more modern version, but so far, I did not see those features we really want to have.

Not sure about you, but these are the features I want to have:

### JEPs

 - [286: Local-Variable Type Inference](https://openjdk.java.net/jeps/286) would help remove noise at declarations.
 - [307: Parallel Full GC for G1](https://openjdk.java.net/jeps/307) could improve performance.
 - [328: Flight Recorder](https://openjdk.java.net/jeps/328) enables better profiling.
 - [361: Switch Expressions](https://openjdk.java.net/jeps/361) improves legibility of switch statements.
 - [386: Alpine Linux Port](https://openjdk.java.net/jeps/386) makes packaging the solver as Docker image (artifact at conference) easier.
 - [377: ZGC: A Scalable Low-Latency Garbage Collector (Production)](https://openjdk.java.net/jeps/377) could improve performance.
 - [378: Text Blocks](https://openjdk.java.net/jeps/378) makes code more readable.
 - [379: Shenandoah: A Low-Pause-Time Garbage Collector (Production)](https://openjdk.java.net/jeps/379) could improve performance.
 - [394: Pattern Matching for instanceof](https://openjdk.java.net/jeps/394) makes code more readable/concise.
 - [395: Records](https://openjdk.java.net/jeps/395)
 
### Overall Performance Improvement
 - [OptaPlanner: How much faster is Java 11?](https://www.optaplanner.org/blog/2019/01/17/HowMuchFasterIsJava11.html)
 - [OptaPlanner: How much faster is Java 17?](https://www.optaplanner.org/blog/2021/09/15/HowMuchFasterIsJava17.html)

### Native Image

We could compile Alpha down to an ELF x86_64 binary with [GraalVM Native Image](https://www.graalvm.org/reference-manual/native-image/). I have done that for [ATLAS](https://github.com/lorenzleutgeb/atlas).

**Comment by madmike200590** [comment]
I agree with @lorenzleutgeb on this one, there are definitely features we want to have from newer java versions - mainly records and flight recorder in my case.

@AntoniusW, in response to your point:

> I think it is an advantage if our software is able to run on "ancient" versions of Java

I'm not so sure if restricting ourselves to old versions as compile targets is the way to go here. The only reason anyone would really have to upgrade their local java setup to the version used by Alpha would be if they want to work on the code (in which case I'm happy with requiring people to do so), whenever people only want to use Alpha, we can provide either versions with bundled JREs (this got more comfortable in JAva 9 with JPMS, see [jlink](https://docs.oracle.com/javase/9/tools/jlink.htm#JSWOR-GUID-CECAC52B-CFEE-46CB-8166-F17A8E9280E9#JSWOR-GUID-CECAC52B-CFEE-46CB-8166-F17A8E9280E9) ) or images created using GraalVM with every release.

**Comment by madmike200590** [comment]
Just to re-kindle the discussion here, I held off on PR #315 back then because I found something like [this survey on java version distribution](https://newrelic.com/resources/report/2022-state-of-java-ecosystem) and was shocked by how scarce applications in Java 17 are. Back then, I thought we couldn't just leave out half the java applications that might want to depend on Alpha. However, since big projects like Jenkins are now finally getting rid of Java 8 support (see [here](https://stackoverflow.com/questions/72812600/does-jenkins-continue-support-java1-8#:~:text=Java%208%20is%20supported%20on,Java%20versions%20are%20not%20supported.) and [here](https://www.jenkins.io/doc/administration/requirements/java/)), I think we should do so too.
I still think it's too early for Java 17 (again, see above survey, I don't think we should require a version with less than 30% distribution), but would like to go to 11.

@AntoniusW and @lorenzleutgeb what are your thoughts? If you're ok with that, I'd go ahead and bump the Java version to 11 in the next couple of weeks.

**Comment by AntoniusW** [comment]
I concur with @madmike200590. Given that a majority of Java applications seems to be on version 11 or newer and some important projects are dropping 8, I think we are fine with dropping 8, too.
Regarding Java 17, it still seems a bit too early for that, since Alpha should be employable also in existing projects and it appears that there are not that many on Java 17 already. This seems to change however, and maybe in a year or so, it makes sense to switch to Java 17.
